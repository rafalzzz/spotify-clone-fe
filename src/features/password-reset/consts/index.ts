export { PASSWORD_RESET_FORM_FIELDS, PASSWORD_RESET_COMPLETE_FORM_FIELDS } from './form-fields';
export {
  PASSWORD_RESET_INITIAL_VALUES,
  PASSWORD_RESET_COMPLETE_INITIAL_VALUES,
} from './initial-values';
export { FORM_FIELD_PLACEHOLDERS } from './form-field-placeholders';
export { FORM_LABELS } from './form-field-labels';
