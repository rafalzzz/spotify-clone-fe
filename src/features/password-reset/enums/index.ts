export { PasswordResetFormKeys } from './password-reset-form-keys';
export { PasswordResetCompleteFormKeys } from './password-reset-complete-form-keys';
