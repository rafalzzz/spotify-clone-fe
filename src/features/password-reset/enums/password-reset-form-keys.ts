export enum PasswordResetFormKeys {
  LOGIN = 'login',
}
