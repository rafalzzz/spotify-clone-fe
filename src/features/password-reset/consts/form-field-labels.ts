import { PasswordResetFormKeys, PasswordResetCompleteFormKeys } from '@/password-reset/enums';

export const FORM_LABELS = {
  [PasswordResetFormKeys.LOGIN]: 'Email address or username',
  [PasswordResetCompleteFormKeys.PASSWORD]: 'New password',
  [PasswordResetCompleteFormKeys.REPEAT_PASSWORD]: 'Repeat new password',
};
