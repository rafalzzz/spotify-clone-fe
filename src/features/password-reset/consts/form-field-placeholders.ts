import { PasswordResetFormKeys, PasswordResetCompleteFormKeys } from '@/password-reset/enums';

import { FORM_LABELS } from './form-field-labels';

export const FORM_FIELD_PLACEHOLDERS = {
  [PasswordResetFormKeys.LOGIN]: FORM_LABELS[PasswordResetFormKeys.LOGIN],
  [PasswordResetCompleteFormKeys.NEW_PASSWORD]: '',
  [PasswordResetCompleteFormKeys.REPEAT_NEW_PASSWORD]: '',
};
