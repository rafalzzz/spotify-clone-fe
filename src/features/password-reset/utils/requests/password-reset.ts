import { TPasswordResetForm } from '@/password-reset/types';

import { customRequest } from '@/utils/custom-request';

import { ENDPOINTS } from '@/consts/endpoints';

export async function passwordReset(requestBody: TPasswordResetForm) {
  const response = customRequest({
    endpoint: ENDPOINTS.PASSWORD_RESET,
    method: 'POST',
    requestBody,
  })
    .then(async (response) => {
      if (response.status === 200) {
        return;
      }

      return await response.json();
    })
    .catch(() => 'Something went wrong');

  return response;
}
