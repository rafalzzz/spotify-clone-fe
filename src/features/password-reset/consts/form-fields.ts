import { repeatNewPasswordValidator } from '@/password-reset/utils/validators/repeat-new-password-validator';

import { EInputType } from '@/enums/input-type';

import { TCustomFormItem } from '@/types/components';

import { passwordValidator, emailOrUsernameValidator } from '@/shared/validators';

import { EPasswordResetCompleteForm, EPasswordResetFormKeys } from '../types';

import { FORM_LABELS } from './form-field-labels';
import { FORM_FIELD_PLACEHOLDERS } from './form-field-placeholders';

export const PASSWORD_RESET_FORM_FIELDS: TCustomFormItem<EPasswordResetFormKeys>[] = [
  {
    type: EInputType.TEXT,
    key: EPasswordResetFormKeys.LOGIN,
    name: EPasswordResetFormKeys.LOGIN,
    label: FORM_LABELS[EPasswordResetFormKeys.LOGIN],
    rules: [{ validator: emailOrUsernameValidator }],
    inputProps: {
      placeholder: FORM_FIELD_PLACEHOLDERS[EPasswordResetFormKeys.LOGIN],
    },
  },
];

export const PASSWORD_RESET_COMPLETE_FORM_FIELDS: TCustomFormItem<EPasswordResetCompleteForm>[] = [
  {
    type: EInputType.PASSWORD,
    key: EPasswordResetCompleteForm.PASSWORD,
    name: EPasswordResetCompleteForm.PASSWORD,
    label: FORM_LABELS[EPasswordResetCompleteForm.PASSWORD],
    rules: [{ validator: passwordValidator }],
    inputProps: {
      placeholder: FORM_FIELD_PLACEHOLDERS[EPasswordResetCompleteForm.PASSWORD],
    },
  },
  {
    type: EInputType.PASSWORD,
    key: EPasswordResetCompleteForm.REPEAT_PASSWORD,
    name: EPasswordResetCompleteForm.REPEAT_PASSWORD,
    label: FORM_LABELS[EPasswordResetCompleteForm.REPEAT_PASSWORD],
    rules: [repeatNewPasswordValidator],
    inputProps: {
      placeholder: FORM_FIELD_PLACEHOLDERS[EPasswordResetCompleteForm.REPEAT_PASSWORD],
    },
  },
];
