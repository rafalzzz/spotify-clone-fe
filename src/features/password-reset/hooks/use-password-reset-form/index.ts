import { useRouter } from 'next/navigation';
import { useMemo, useState } from 'react';

import { TPasswordResetForm, TUsePasswordResetFormProps } from '@/password-reset/types';
import { passwordReset } from '@/password-reset/utils/requests/password-reset';

import { TCustomButton } from '@/types/components';
import { THookForm } from '@/types/components';

export const usePasswordResetForm = ({ displayError }: THookForm): TUsePasswordResetFormProps => {
  const [isLoading, setIsLoading] = useState(false);

  const { push } = useRouter();

  const onFinish = async (values: TPasswordResetForm) => {
    setIsLoading(true);

    try {
      const response = await passwordReset(values);

      if (!response) {
        push('/');
      }

      if (response) {
        displayError(response as string);
      }
    } catch (error: unknown) {
      displayError(error as string);
    } finally {
      setIsLoading(false);
    }
  };

  const submitButton: TCustomButton = useMemo(
    () => ({
      htmlType: 'submit',
      shape: 'round',
      text: 'Send',
      disabled: isLoading,
      testId: 'submit-button',
      className: 'login-button',
    }),
    [isLoading],
  );

  return { submitButton, onFinish };
};
