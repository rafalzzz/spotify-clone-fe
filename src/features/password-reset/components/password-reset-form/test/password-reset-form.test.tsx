import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

import { FORM_FIELD_PLACEHOLDERS } from '@/password-reset/consts';
import { PasswordResetFormKeys } from '@/password-reset/enums/password-reset-form-keys';
import { usePasswordResetForm } from '@/password-reset/hooks/use-password-reset-form';

import { emailOrUsernameValidator } from '@/validators/email-or-username-validator';

import { InputType } from '@/enums/input-type';

import { PasswordResetForm } from '..';

jest.mock('@/password-reset/hooks/use-password-reset-form');

jest.mock('@/password-reset/utils/requests/password-reset', () => ({
  passwordReset: jest.fn(() => Promise.resolve({ data: {} })),
}));

jest.mock('@/validators/email-or-username-validator', () => ({
  emailOrUsernameValidator: jest
    .fn()
    .mockImplementation((getFieldValue) => () => Promise.resolve()),
}));

const renderPasswordResetForm = () => render(<PasswordResetForm />);

const onFinishMock = jest.fn();

describe('PasswordResetForm', () => {
  beforeEach(() => {
    (usePasswordResetForm as jest.Mock).mockReturnValue({
      submitButton: {
        key: 1,
        type: 'primary',
        htmlType: 'submit',
        text: 'Send',
        disabled: false,
        testId: 'submit-button',
      },
      onFinish: onFinishMock,
    });
  });

  it('render component without error', () => {
    const screen = renderPasswordResetForm();
    expect(screen).toMatchSnapshot();
  });

  it('should call validators and not call mocked action when form fields are empty', async () => {
    const { queryByTestId } = renderPasswordResetForm();
    const submitButton = queryByTestId('submit-button');
    expect(submitButton).toBeInTheDocument();

    await userEvent.click(submitButton as Element);

    // test validators calls
    expect(emailOrUsernameValidator).toHaveBeenCalled();

    // test mocked function call
    expect(onFinishMock).not.toHaveBeenCalled();
  });

  describe('should call validator when input value change', () => {
    const formFieldsWithValidators = [
      {
        type: InputType.TEXT,
        key: PasswordResetFormKeys.LOGIN,
        placeholder: FORM_FIELD_PLACEHOLDERS[PasswordResetFormKeys.LOGIN],
        mockedValue: 'Test',
        validator: emailOrUsernameValidator,
      },
    ];

    formFieldsWithValidators.forEach(({ type, key, validator, ...restProps }) => {
      it(key, async () => {
        const { queryByPlaceholderText } = renderPasswordResetForm();

        const { mockedValue, placeholder } = restProps;
        const input = queryByPlaceholderText(placeholder);
        expect(input).toBeInTheDocument();

        await userEvent.type(input as Element, mockedValue);

        expect(validator).toHaveBeenCalled();
        return;
      });
    });
  });
});
