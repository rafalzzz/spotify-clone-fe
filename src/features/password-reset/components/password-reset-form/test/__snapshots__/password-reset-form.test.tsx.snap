// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordResetForm render component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-vertical ant-form-large css-dev-only-do-not-override-12upa3x password-reset-form"
        id="password-reset"
      >
        <div
          class="ant-form-item form-item css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-12upa3x"
            >
              <label
                class=""
                for="password-reset_login"
                title=""
              >
                <span
                  class="form-item__label"
                >
                  Email address or username
                </span>
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input ant-input-lg css-dev-only-do-not-override-12upa3x"
                    data-testid="input-type-text"
                    id="password-reset_login"
                    placeholder="Email address or username"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item submit-button css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-primary ant-btn-lg"
                    data-testid="submit-button"
                    role="button"
                    type="submit"
                  >
                    <span>
                      Send
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-vertical ant-form-large css-dev-only-do-not-override-12upa3x password-reset-form"
      id="password-reset"
    >
      <div
        class="ant-form-item form-item css-dev-only-do-not-override-12upa3x"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-12upa3x"
          >
            <label
              class=""
              for="password-reset_login"
              title=""
            >
              <span
                class="form-item__label"
              >
                Email address or username
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input ant-input-lg css-dev-only-do-not-override-12upa3x"
                  data-testid="input-type-text"
                  id="password-reset_login"
                  placeholder="Email address or username"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item submit-button css-dev-only-do-not-override-12upa3x"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-primary ant-btn-lg"
                  data-testid="submit-button"
                  role="button"
                  type="submit"
                >
                  <span>
                    Send
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
